src/Microsoft.NetCore.Analyzers/Core/Security/DoNotCallDangerousMethodsInDeserialization.cs

```diff
@@ -0,0 +1,242 @@
+
+                    var dangerousMethodSymbols = dangerousMethodSymbolsBuilder.ToImmutable();
+
+                    if (dangerousMethodSymbols.Length == 0)
+                    {
+                        return;
+                    }
+
+                    var attributeTypeSymbolsBuilder = ImmutableArray.CreateBuilder<INamedTypeSymbol>();
+                    var onDeserializingAttributeTypeSymbol = WellKnownTypes.OnDeserializingAttribute(compilation);
+
+                    if (onDeserializingAttributeTypeSymbol != null)
+                    {
+                        attributeTypeSymbolsBuilder.Add(onDeserializingAttributeTypeSymbol);
+                    }
+
+                    var onDeserializedAttributeTypeSymbol = WellKnownTypes.OnDeserializedAttribute(compilation);
+
+                    if (onDeserializedAttributeTypeSymbol != null)
+                    {
+                        attributeTypeSymbolsBuilder.Add(onDeserializedAttributeTypeSymbol);
+                    }
+
+                    var attributeTypeSymbols = attributeTypeSymbolsBuilder.ToImmutable();
+                    var streamingContextTypeSymbol = WellKnownTypes.StreamingContext(compilation);
+                    var IDeserializationCallbackTypeSymbol = WellKnownTypes.IDeserializationCallback(compilation);
+                    // A dictionary from method symbol to set of methods invoked by it directly.
+                    // The bool value in the sub ConcurrentDictionary is not used, use ConcurrentDictionary rather than HashSet just for the concurrency security.
+                    var callGraph = new ConcurrentDictionary<IMethodSymbol, ConcurrentDictionary<IMethodSymbol, bool>>();
+
+                    compilationStartAnalysisContext.RegisterOperationBlockStartAction(
+                        (OperationBlockStartAnalysisContext operationBlockStartAnalysisContext) =>
+                        {
+                            var owningSymbol = operationBlockStartAnalysisContext.OwningSymbol;
+
+                            if (owningSymbol.Kind != SymbolKind.Method)
+                            {
+                                return;
+                            }
+
+                            var methodSymbol = (IMethodSymbol)owningSymbol;
+                            var classSymbol = methodSymbol.ContainingType;
+
+                            if (!classSymbol.HasAttribute(serializableAttributeTypeSymbol))
+                            {
+                                return;
+                            }
+
+                            var calledMethods = new ConcurrentDictionary<IMethodSymbol, bool>();
+                            callGraph.TryAdd(methodSymbol, calledMethods);
+
+                            operationBlockStartAnalysisContext.RegisterOperationAction(operationContext =>
+                            {
+                                callGraph[methodSymbol].TryAdd((operationContext.Operation as IInvocationOperation).TargetMethod, true);
```

Comment:
I believe you can directly add to `calledMethods` here to avoid additional concurrent accesses into `callGraph` data structure: `calledMethods.TryAdd((IInvocationOperation)operationContext.Operation).TargetMethod, true);`